### PR TITLE
Initial Commit

### PR DESCRIPTION
**Pull request by bdarfler**

### CODE REVIEW

**Comment by bdarfler** [comment]
@jsolmon what do you think?

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+    git checkout -b <my_branch_name>
+
+Code and document your new feature:
+    
+    We appreciate well organized commits, commit messages
+
+Push your changes up to GitHub:
+    
+    git push -u origin <my_branch_name>
```
Comment (jsolmon) [inline CONTRIBUTING.md:17]: I don't think users outside of Localytics can push new branches to the repo. We should use the fork workflow instead.

Comment (bdarfler) [inline CONTRIBUTING.md:17]: I added a line at the time telling users to fork and clone the repo first.

Comment (jsolmon) [inline CONTRIBUTING.md:17]: :+1:

**Comment by jsolmon** [comment]
:shipit:

File: src/main/scala/com/localytics/sbt/s3/S3ProxyTasks.scala [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      if (!downloadDir.exists()) {
+        streamz.log.info(s"Creating S3Proxy directory $downloadDir")
+        downloadDir.mkdirs()
+      }
+      if (!outputFile.exists()) {
+        streamz.log.info(s"Downloading S3Proxy from [$url] to [${outputFile.getAbsolutePath}]")
+        (new URL(url) #> outputFile).!!
+      }
+      if (!outputFile.exists()) {
+        sys.error(s"Cannot to find S3Proxy file at [${outputFile.getAbsolutePath}]")
```
Comment (jsolmon) [inline src/main/scala/com/localytics/sbt/s3/S3ProxyTasks.scala]: Typo - need to take out the `to` in this error

File: src/main/scala/com/localytics/sbt/s3/S3ProxyTasks.scala [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        waitForS3Proxy(port, (s: String) => streamz.log.info(s))
+      }
+      if (extractS3ProxyPid("jps".!!).isEmpty) {
+        sys.error(s"Cannot find S3Proxy PID")
+      }
+  }
+
+  def stopS3ProxyTask = (streams, s3ProxyDataDir, s3ProxyCleanAfterStop) map {
+    case (streamz, dataDir, clean) =>
+      extractS3ProxyPid("jps".!!) match {
```
Comment (jsolmon) [inline src/main/scala/com/localytics/sbt/s3/S3ProxyTasks.scala:60]: Do you need to use `jps -l` here? When I run `jps` on my machine I don't get the name of the jar.

```
~/src/local-backend/$ jps
83826 jar
83928 Jps
78462 MainGenericRunner
```

vs

```
~/src/local-backend/$ jps -l
84448 sun.tools.jps.Jps
83826 elasticmq-server-0.8.12.jar
78462 scala.tools.nsc.MainGenericRunner
```

File: src/main/scala/com/localytics/sbt/s3/S3ProxyUtils.scala [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      val socket = new java.net.Socket("localhost", port)
+      socket.close()
+    }.isSuccess
+  }
+
+  def waitForS3Proxy(port: Int, infoPrintFunc: String => Unit): Unit = {
+    var continue = false
+    while (!continue) {
+      continue = true
+      Try {
```
Comment (jsolmon) [inline src/main/scala/com/localytics/sbt/s3/S3ProxyUtils.scala]: It's a little cleaner to use `isS3ProxyRunning` in your while. Avoids the duplicate code and var.

File: src/main/scala/com/localytics/sbt/s3/S3ProxyUtils.scala [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        val socket = new java.net.Socket("localhost", port)
+        socket.close()
+      }.recover {
+        case e: Exception =>
+          infoPrintFunc(s"Waiting for S3Proxy to boot on port $port")
+          Thread.sleep(500)
+          continue = false
+      }
```
Comment (jsolmon) [inline src/main/scala/com/localytics/sbt/s3/S3ProxyUtils.scala]: Also, shouldn't this time out at some point?

File: src/main/scala/com/localytics/sbt/s3/S3ProxyUtils.scala (hunk 4 above)
Comment (bdarfler) [inline src/main/scala/com/localytics/sbt/s3/S3ProxyUtils.scala]: Glad you are going over this more carefully!

Comment (bdarfler) [inline src/main/scala/com/localytics/sbt/s3/S3ProxyUtils.scala]: I went a step further and just inlined the method. What do you think?

Comment (jsolmon) [inline src/main/scala/com/localytics/sbt/s3/S3ProxyUtils.scala]: :+1:

**Pull request merged**